Add support for CIFS Mounts on Windows

blueelvis opened the pull request:
<!-- 🎉 Thank you for contributing to minikube! 🎉 Here are some hints to get your PR merged faster:

1. Your PR title will be included in the release notes, choose it carefully
2. If the PR fixes an issue, add "fixes #<issue number>" to the description.
3. If the PR is a user interface change, please include a "before" and "after" example.
4. If the PR is a large design change, please include an enhancement proposal:
https://github.com/kubernetes/minikube/tree/master/enhancements
-->


Related to #1197 . Adding CIFS support should allow mounting Azure Files with minimal changes.

blueelvis commented:
@tstromberg @sharifelgamal -- ping. I have moved all of the commands which I foresee to be useful in minikube from docker-machine over here. Let me know how this looks. Ofcourse, this is experimental!

tstromberg commented:
Travis:

```
cmd/minikube/cmd/mount.go:217: undefined: hyperv.ConfigureHostMount
cmd/minikube/cmd/mount.go:261: undefined: hyperv.IsWindowsAdministrator
cmd/minikube/cmd/mount.go:267: undefined: hyperv.GetCurrentWindowsUser
```

tstromberg commented on pkg/minikube/drivers/hyperv/windows.go:
```diff
@@ -0,0 +1,137 @@
+	return resp[0] == "True"
+}
+
+func IsWindowsAdministrator() (bool, error) {
```
Consider whether you need both `bool` and `error`, I notice that elsewhere you only refer to one of the return values. Perhaps this should have a signature like:

`func VerifyAdminRole() error`

tstromberg commented on cmd/minikube/cmd/mount.go:
```diff
@@ -225,3 +251,45 @@ func getPort() (int, error) {
+	// Check if the Credential exists in the credential store.
+	var credentialName = "minikube"
+	var password = ""
+	cred, err := wincred.GetGenericCredential(credentialName)
```
Any calls to `wincred` should be constrained to files with a `// +build windows` rule.

tstromberg commented on cmd/minikube/cmd/mount.go:
```diff
@@ -225,3 +251,45 @@ func getPort() (int, error) {
+	if err != nil {
+		return err
+	}
+	user = strings.Replace(user,(hostname + "\\"), "",1)
```
Can you add a comment about what this is doing? It isn't clear to me.

tstromberg commented on cmd/minikube/cmd/mount.go:
```diff
@@ -225,3 +251,45 @@ func getPort() (int, error) {
+		out.T(out.Notice,"Credential {{.credential}} was found in the Windows Credential Store. Using that...",out.V{"credential":credentialName})
+		password = string(cred.CredentialBlob)
+	} else {
+		out.T(out.Enabling,"Please Type in the password for the user - [{{.username}}]",out.V{"username":user})
```
Type does not have to be capitalized.

You should also mention what this password is going to be used for.

tstromberg commented on cmd/minikube/cmd/mount.go:
```diff
@@ -225,3 +251,45 @@ func getPort() (int, error) {
+
+		cred := wincred.NewGenericCredential(credentialName)
+		cred.CredentialBlob = []byte(inputPassword)
+		wincrederr := cred.Write()
```
this can just be `err`

tstromberg commented on cmd/minikube/cmd/mount.go:
```diff
@@ -225,3 +251,45 @@ func getPort() (int, error) {
+		password = string(cred.CredentialBlob)
+	} else {
+		out.T(out.Enabling,"Please Type in the password for the user - [{{.username}}]",out.V{"username":user})
+		inputPassword, _ := terminal.ReadPassword(int(os.Stdin.Fd()))
```
Please don't throw away `err`.

blueelvis replied:
My bad on this. Fixed.

tstromberg commented on cmd/minikube/cmd/mount.go:
```diff
@@ -225,3 +251,45 @@ func getPort() (int, error) {
+	cred, err := wincred.GetGenericCredential(credentialName)
+	if err == nil {
+		out.T(out.Notice,"Credential {{.credential}} was found in the Windows Credential Store. Using that...",out.V{"credential":credentialName})
+		password = string(cred.CredentialBlob)
```
Consider isolating the function that retrieves/saves the password into it's own function. That way there is a clearer flow of where the password is coming from.

blueelvis replied:
I have created 2 separate functions i.e. `getMountPassword()` & `saveMountPassword()`

tstromberg commented on cmd/minikube/cmd/mount.go:
```diff
@@ -225,3 +251,45 @@ func getPort() (int, error) {
+	mountCmd := fmt.Sprintf("sudo mkdir -p %s && sudo mount.cifs //%s/%s %s -o username=%s,password=%s,domain=%s",vmDestinationPath, hostname, shareName, vmDestinationPath, user, password, hostname)
+	err = cluster.MountCifs(runner, vmDestinationPath, mountCmd)
+	if err != nil {
+		out.ErrT(out.FailureType, "Failed mounting: {{.error}}", out.V{"error": err})
```
Either display the error, or return it, but not both. Doing so results in duplicated error messages being shown.

tstromberg commented on pkg/minikube/cluster/mount.go:
```diff
@@ -69,6 +69,22 @@ func Mount(r mountRunner, source string, target string, c *MountConfig) error {
+	out, err := r.CombinedOutput(mountCommand)
+	if err != nil {
+		//glog.Infof("%s failed: err=%s, output: %q", mountCommand, err, out)
+		glog.Infof("Mounting failed: err=%s, output: %q", err, out)
```
This should return err.

blueelvis commented:
Thanks for the detailed review @tstromberg ! Working on this now.

blueelvis commented on pkg/minikube/drivers/hyperv/driver.go:
```diff
@@ -51,3 +51,33 @@ func createHypervHost(config cfg.MachineConfig) interface{} {
 	return d
 }
+
+func ConfigureHostMount(shareName string, hostPath string) error  {
```
It took me quite some time for understanding how interfaces are working in golang... I am used to "implements". This is complete.

blueelvis commented on pkg/minikube/drivers/hyperv/windows.go:
```diff
@@ -0,0 +1,137 @@
+
+func powershellCmdOut(args ...string) (string, error) {
+	args = append([]string{"-NoProfile", "-NonInteractive"}, args...)
+	cmd := exec.Command(powershell, args...)
```
I moved the logic from`init` to this section. If Powershell is not available, it should exit first and not even try running the command.

blueelvis commented on pkg/minikube/drivers/hyperv/windows.go:
```diff
@@ -0,0 +1,137 @@
+func powershellCmdOut(args ...string) (string, error) {
+	args = append([]string{"-NoProfile", "-NonInteractive"}, args...)
+	cmd := exec.Command(powershell, args...)
+	glog.Infof("[executing ==>] : %v %v", powershell, strings.Join(args, " "))
```
I had copied this from the Docker Machine library :) . Removed and edited.

blueelvis commented on pkg/minikube/drivers/hyperv/windows.go:
```diff
@@ -0,0 +1,137 @@
+	return err
+}
+
+func parseLines(stdout string) []string {
```
I see the usage of `ParseLines` function. This is not just trimming space but rather provides the entire output line by line in a way it can be iterated over. I would prefer keeping this function as it is. What do you think?

blueelvis commented on pkg/minikube/drivers/hyperv/windows.go:
```diff
@@ -0,0 +1,137 @@
+
+// This function is used to get the full name of the current user trying to run. It will be DOMAIN\Username or MACHINE_NAME\Username
+// TODO - Check if CIFS shares can be used by people who have domain accounts and are local admins on their machines.
+func GetCurrentWindowsUser() (string, error) {
```
Changed. The `user.Current()` returns a lot of things which are not required. For example,
it returned this string -- `&{S-1-5-21-3668668534-1506658371-4068906332-1001 S-1-5-21-3668668534-1506658371-4068906332-1001 DESKTOP\blue John Doe C:\Users\blue}`

Although I agree that it is part of the current user's information, I wonder if this should be reported to golang so that this can be enhanced for Windows? What do you think?

blueelvis commented on cmd/minikube/cmd/mount.go:
```diff
@@ -225,3 +251,45 @@ func getPort() (int, error) {
 	return l.Addr().(*net.TCPAddr).Port, nil
 }
+
+func enableCifsShare(hostShareName string, vmDestinationPath string, runner command.Runner) (error) {
```
Removed this function completely since we used the interface approach.

blueelvis commented:
@tstromberg - Apologies for these delayed changes. Have been travelling :( 

How does this next pass look to you? I have addressed the review comments (Again thanks for the fantastic review comments!) and have tested on my machine and this seems to be working. There are quite a lot of improvements for this in pipe once this is merged :)

blueelvis commented:
Fixed the boilerplate warning. Thanks @afbjorklund!

tstromberg commented:
/ok-to-test

tstromberg commented on cmd/minikube/cmd/mount.go:
```diff
@@ -174,36 +188,63 @@ var mountCmd = &cobra.Command{
+			out.T(out.Notice, "NOTE: This process must stay alive for the mount to be accessible ...")
+			wg.Wait()
+		} else if cfg.Type == cifs {
+			if host.Driver.DriverName() == driver.HyperV {
```
I don't think believe this has to be HyperV specific.

tstromberg commented on cmd/minikube/cmd/mount.go:
```diff
@@ -174,36 +188,63 @@ var mountCmd = &cobra.Command{
+				}
+				err = cifsMounter.Mount(runner)
+				if err != nil {
+					exit.WithError("cifs mount ", err)
```
remove trailing space

tstromberg commented on pkg/minikube/mount/cifs_windows.go:
```diff
@@ -0,0 +1,159 @@
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/ssh/terminal"
+	"k8s.io/minikube/pkg/minikube/drivers/hyperv"
```
Copy or move the hyperv functions into this package instead of making hyperv a dependency of mount.

blueelvis replied:
I am not clear on this. The Hyper-V related functions should be separate and seem to be in the correct place? The mount dependency on Hyper-V is only when it is built against Hyper-V. Could you please rephrase?

tstromberg commented on pkg/minikube/mount/cifs_windows.go:
```diff
@@ -0,0 +1,159 @@
+	"strings"
+)
+
+var credentialName = "minikube"
```
Make this a const if possible.

tstromberg commented on pkg/minikube/mount/cifs_windows.go:
```diff
@@ -0,0 +1,159 @@
+}
+
+func (m *Cifs) Mount(r mountRunner) error {
+	hostname, err := os.Hostname()
```
Depending on a specific DNS configuration adds room for failure. Pass in the result of `cluster.GetVMHostIP()`, as the 9p mount does.

This should probably reuse `Config.HostPath` instead of ignoring it.

tstromberg commented on pkg/minikube/mount/cifs_windows.go:
```diff
@@ -0,0 +1,159 @@
+		}
+	}
+
+	mountCmd := fmt.Sprintf("sudo mkdir -p %s && sudo mount.cifs //%s/%s %s -o username=%s,password=%s,domain=%s",m.VMDestinationPath, hostname, m.HostShareName, m.VMDestinationPath, username, mountPassword , domain)
```
Use `credentials` instead of `password`, to avoid leaking the password onto the command-line.

I'd also recommend deleting the credentials file once mount succeeds.

tstromberg commented on pkg/minikube/mount/cifs_windows.go:
```diff
@@ -0,0 +1,159 @@
+}
+
+func (m *Cifs) Unmount(r mountRunner) error {
+	// Unmount the minikube destination path
```
Remove this comment - it does not add helpful context.

tstromberg commented on pkg/minikube/mount/cifs_windows.go:
```diff
@@ -0,0 +1,159 @@
+	if err == nil {
+		glog.Infof("Credential %v was found in the Windows Credential Store. Using that...",credentialName)
+		return string(cred.CredentialBlob), nil
+	} else {
```
Handle and return errors first rather than using an else statement: this way the exceptional path is highlighted, and the non-exceptional one is not indented.

tstromberg commented on pkg/minikube/mount/mount.go:
```diff
@@ -0,0 +1,121 @@
+)
+
+var (
+	ErrNotImplemented = errors.New("This type of Mount is not available on this system")
```
no need to capitalize mount.

tstromberg commented on pkg/minikube/mount/mount.go:
```diff
@@ -0,0 +1,121 @@
+	Options map[string]string
+}
+
+type Cifs struct {
```
- The CIFS struct should be named `CIFS`: https://github.com/golang/go/wiki/CodeReviewComments#initialisms
- It should be moved to `cifs.go`
- Instead of duplicating the contents of `Config`, add a private `config` accessor:

```
type CIFS struct {
  config Config
}
```

tstromberg commented:
Any progress?

priyawadhwa commented:
Hey @blueelvis I'm going to close this PR for now as some time has passed, but please reopen once comments are addressed. Thank you!
